Hw06

tikalet opened the pull request.

tikalet committed:
add homework 06

tikalet committed:
add comment crud

tikalet committed:
add test for comment

tikalet committed:
fix checkstyle

vlad-iv commented on hw06/src/main/java/ru/otus/hw/models/Book.java:
```diff
@@ -0,0 +1,45 @@
+import lombok.NoArgsConstructor;
+import lombok.ToString;
+
+@Data
```
Можно удалить, заменив на Getter / Setter
т.е. equals и hashCode не совместимы с Entity
могут загружать Lazy зависимости

vlad-iv commented on hw06/src/main/java/ru/otus/hw/models/BookComment.java:
```diff
@@ -0,0 +1,39 @@
+@NoArgsConstructor
+@Entity
+@Table(name = "book_comments")
+@NamedEntityGraph(name = "book-comment-entity-graph", attributeNodes = {@NamedAttributeNode("book")})
```
(optional) можно удалить, т.к. книгу загружать не потребуется, мы по книге (загруженной) получаем ее комментарии

vlad-iv commented on hw06/src/main/java/ru/otus/hw/models/BookComment.java:
```diff
@@ -0,0 +1,39 @@
+@Entity
+@Table(name = "book_comments")
+@NamedEntityGraph(name = "book-comment-entity-graph", attributeNodes = {@NamedAttributeNode("book")})
+public class BookComment {
```
Можно переименовать в Comment

vlad-iv commented on hw06/src/main/java/ru/otus/hw/repositories/JpaBookRepository.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public void deleteById(long id) {
+        Optional<Book> book = findById(id);
```
предлагаю явно вызвать метод find
при реализации публичных методов, рекомендуется избегать вызова других публичных методов
(в нашем случаем, на не нужно применять граф и другую логику из публичного метода)

vlad-iv commented on hw06/src/main/java/ru/otus/hw/repositories/JpaBookRepository.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public Optional<Book> findById(long id) {
+        EntityGraph<?> entityGraph = entityManager.getEntityGraph("book-entity-graph");
+        TypedQuery<Book> query = entityManager.createQuery("SELECT b FROM Book b WHERE b.id = :id", Book.class);
```
можно заменить на метод find

vlad-iv commented on hw06/src/test/java/ru/otus/hw/repositories/JpaBookRepositoryTest.java:
```diff
@@ -0,0 +1,97 @@
+@DisplayName("Репозиторий на основе JPA для работы с книгами ")
+@DataJpaTest
+@Import({JpaBookRepository.class})
+class JpaBookRepositoryTest {
```
Осталось Написать интеграционные тесты сервисов книг и комментариев, которые будут проверять работу с БД. Транзакционность в этих тестах должна быть отключена, чтобы не влияла на транзакции в сервисах. Проверить, что доступ к связям, которые используются снаружи сервисов не вызывают LazyInitialzationException. Не забыть учесть кэширование контекста в тестах

tikalet committed:
fix

- replace @Data to @Getter && @Setter
- delete DTO
- delete load book for comment
- use "find" for load comment entity
- change delete method

tikalet committed:
rename BookComment to Comment

tikalet committed:
add integration tests for comment

tikalet committed:
add integration tests for book

vlad-iv commented on hw06/src/main/java/ru/otus/hw/repositories/JpaBookRepository.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Override
+    public Optional<Book> findById(long id) {
+        Book comment = entityManager.find(Book.class, id);
```
осталось добавить хинт  с book-entity-graph

vlad-iv commented on hw06/src/main/java/ru/otus/hw/services/BookServiceImpl.java:
```diff
@@ -0,0 +1,68 @@
+    @Override
+    @Transactional
+    public Book update(long id, String title, long authorId, long genreId) {
+        bookRepository.findById(id).
```
Предлагаю сохранить в переменную 
После через сеттеры задать новые значения полям

vlad-iv commented on hw06/src/main/java/ru/otus/hw/services/CommentServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    @Transactional
+    public Comment update(long id, String comment, long bookId) {
+        commentRepository.findById(id).
```
Можно удалить параметр bookId
сохранить найденный комментарий в переменную 
и через сеттер задать новый текст

tikalet committed:
fix

- add entityGraph in find for book
- change update comment, book
- delete dto
- delete check size for list method

tikalet committed:
fix check style for tests
